lib/ssl/src/tls_sender.erl

```diff
@@ -302,39 +320,54 @@ connection({call, From}, {dist_handshake_complete, _Node, DHandle},
-    SecurityParams = maps:get(security_parameters, CurrentWrite),
-    ApplicationTrafficSecret =
-        SecurityParams#security_parameters.application_traffic_secret,
-    hibernate_after(connection, Data,
-                    [{reply, From, {ok, ApplicationTrafficSecret, N}}]);
+
 connection(internal, {application_packets, From, Data}, StateData) ->
     send_application_data(Data, From, connection, StateData);
-connection(internal, {post_handshake_data, From, HSData}, StateData) ->
-    send_post_handshake_data(HSData, From, connection, StateData);
-connection(cast, #alert{} = Alert, StateData0) ->
-    StateData = send_tls_alert(Alert, StateData0),
-    {next_state, connection, StateData};
-connection(cast, {new_write, WritesState, Version}, 
+connection(internal, {post_handshake_data, From, HSData}, #data{buff = Buff} = StateData) ->
+     case Buff of
+         undefined ->
+             send_post_handshake_data(HSData, From, connection, StateData);
+         Async ->
+             {next_state, async_wait, StateData#data{buff = Async#async{low = 0}}, [postpone]}
+     end;
+
+connection(cast, #alert{} = Alert,  #data{buff = Buff} = StateData0) ->
+     case Buff of
+         undefined ->
+             StateData = send_tls_alert(Alert, StateData0),
+             {next_state, connection, StateData};
+         Async ->
+             {next_state, async_wait, StateData0#data{buff = Async#async{low = 0}}, [postpone]}
+     end;
+connection(cast, {new_write, WritesState, Version},
            #data{connection_states = ConnectionStates, env = Env} = StateData) ->
+    CW = maps:remove(aead_handle, WritesState),
     hibernate_after(connection,
-                    StateData#data{connection_states =
-                                       ConnectionStates#{current_write => maps:remove(aead_handle, WritesState)},
-                                   env =
-                                       Env#env{negotiated_version = Version}}, []);
+                    StateData#data{connection_states = ConnectionStates#{current_write => CW},
+                                   env = Env#env{negotiated_version = Version}}, []);
 %%
 connection(info, dist_data,
            #data{env = #env{dist_handle = DHandle}} = StateData) ->
       case dist_data(DHandle) of
           [] -> hibernate_after(connection, StateData, []);
           Data -> send_application_data(Data, dist_data, connection, StateData)
       end;
-connection(info, tick, StateData) ->  
+connection(info, {'$socket', Socket, select, Handle},
+           #data{env = #env{transport_cb = Transport}} = StateData) ->
+    do_async_send(Transport, Socket, Handle, connection, ok, StateData);
+connection(info, {'$socket', Socket, completion, {Handle, Status}},
+           #data{env = #env{transport_cb = Transport}} = StateData) ->
+    do_async_send(Transport, Socket, Handle, connection, Status, StateData);
+connection(info, tick, #data{buff = Buff} = StateData) ->
```

Comment:
Should we perhaps clarify that tick clauses are  Erlang distribution clauses?! This was not done by old code either, but I think it could help easy of understanding.